TASK: Rework CacheSegmentParser to a recursive pattern

kitsunet opened the pull request:
Rewrite the internals of the CacheSegmentParser to use a recursive
approach which should safe some memory and makes the code more readable.
Also refactors common logic for fetching tokens and content to methods
to avoid code duplication.

Additionally adds calculation of uncached segments in the currently parsed
content to make solid cache entries (full page cache) possible.

daniellienert commented on TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/CacheSegmentParser.php:
```diff
@@ -21,134 +22,227 @@ class CacheSegmentParser
+    public function __construct($content, $randomCacheMarker = '')
     {
+        $this->randomCacheMarker = $randomCacheMarker;
+        $this->uncachedPartCount = 0;
```
Is this needed, as it is also initialized in the property definition?

daniellienert commented on TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/CacheSegmentParser.php:
```diff
@@ -21,134 +22,227 @@ class CacheSegmentParser
+     * @param integer $currentPosition
+     * @return array
+     * @throws Exception
+     * @throws \Exception
```
Can be removed

daniellienert commented on TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/CacheSegmentParser.php:
```diff
@@ -21,134 +22,227 @@ class CacheSegmentParser
+    {
+        $nextStartPosition = $this->calculateNextTokenPosition($currentPosition, ContentCache::CACHE_SEGMENT_START_TOKEN);
+        if ($nextStartPosition !== $currentPosition) {
+            throw new Exception(sprintf('The current position (%d) is not the start of a segment, next start position $d', $currentPosition, $nextStartPosition), 1472464124);
```
"position $d'" should be "position %d'" ?

daniellienert commented on TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/CacheSegmentParser.php:
```diff
@@ -21,134 +22,227 @@ class CacheSegmentParser
+     * @param array $segmentData
+     * @return array
+     */
+    protected function extractContentAndSubSegments($currentPosition, $segmentData)
```
typehint "array $segementData" missingß

daniellienert commented on TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/CacheSegmentParser.php:
```diff
@@ -21,134 +22,227 @@ class CacheSegmentParser
+        return $segmentData;
+    }
+
+    protected function reduceSegmentDataToCacheRelevantInformation(array $segmentData)
```
Method comment missing

kitsunet commented:
Fixed comments and logical error that the test exposed. Lets see...

hlubek commented:
@kitsunet Could we check the performance of the substitution of `strpos` with `Functions::strpos` (uses `mb_strpos`)? AFAIK `mb_strpos` is much slower here and not needed.

Otherwise I'm fine to rework this as long as the tests are happy and we don't cause major performance regressions ;)

kitsunet commented:
I can compare, we can also go back to strpos, it's not really important for the refactoring obviously.

hlubek commented on TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/CacheSegmentParser.php:
```diff
@@ -21,134 +22,241 @@ class CacheSegmentParser
+            $segmentData['context'] = $contextOrMetadata;
+            $this->uncachedPartCount++;
+        } elseif (strpos($identifier, 'evalCached=') === 0) {
+            $segmentData['type'] = ContentCache::SEGMENT_TYPE_DYNAMICCACHED;
```
I would expect that `evalCached` also increments `uncachedPartCount` because it also prevents a "solid" cache entry (like static publishing).

dfeyer replied:
Done in https://github.com/neos/neos-development-collection/pull/683/commits/a6c3dacb5a79e2723c158fc06fa06403d7b3cb17

hlubek commented:
Alright, so I did some benchmarking...

I repeated the 19Kb output of the demo site features page 10 times and executed the different cache segment parsers with PHPBench (nice tool by the way).

As can be seen the unicode functions _drastically_ reduce the performance, but using plain `strpos` performs :

```
+-------------------------+-----------------------+--------+--------+------+-----+------------+-------------+-------------+-------------+-------------+---------+--------+---------+
| benchmark               | subject               | groups | params | revs | its | mem_peak   | best        | mean        | mode        | worst       | stdev   | rstdev | diff    |
+-------------------------+-----------------------+--------+--------+------+-----+------------+-------------+-------------+-------------+-------------+---------+--------+---------+
| CacheSegmentParserBench | benchCurrent          |        | []     | 10   | 1   | 2,493,600b | 1.784ms     | 1.784ms     | 1.784ms     | 1.784ms     | 0.000ms | 0.00%  | 0.00%   |
| CacheSegmentParserBench | benchRefactored       |        | []     | 10   | 1   | 2,592,432b | 1,185.504ms | 1,185.504ms | 1,185.504ms | 1,185.504ms | 0.000ms | 0.00%  | +99.85% |
| CacheSegmentParserBench | benchRefactoredStrpos |        | []     | 10   | 1   | 2,562,424b | 4.780ms     | 4.780ms     | 4.780ms     | 4.780ms     | 0.000ms | 0.00%  | +62.67% |
+-------------------------+-----------------------+--------+--------+------+-----+------------+-------------+-------------+-------------+-------------+---------+--------+---------+
```

So basically the new implementations needs more than twice the time which reduces performance when parsing large outputs (which happens on the first hit by the user). I'm not fully happy to degrade the performance here, but it should not matter too much since this is not the bottleneck when rendering a larger page.

dfeyer commented:
@hlubek in https://github.com/neos/neos-development-collection/pull/683/commits/2fefc96f5f4b3ec3947298caf6f010a3fc4212fd i use the PHP strpos, but I have exception "invalid cache segment" ... so I thinks we need the mutibyte version here ?

hlubek commented:
@dfeyer Regarding the usage of `strpos` and exceptions: IMHO it shouldn't matter for the parsing of the token. Can you reproduce that error so we can add another test case for it?

kitsunet commented:
ok, rebased to current master, I squashed all commits because I almost had to apply them by hand...

I also found that removing the unicode functions breaks tests, so I think we need them.

kitsunet commented:
Ok, repushed again, I found a bug within the parser related to substr which the unicode functions fixed indirectly. Now without unicode functions and some minimal speed optimizations.

Should be good to go.

kitsunet commented:
This is still good for review :)

dfeyer commented:
Code wise it look good, I need to test it a bit more, I can do that today. 

@hlubek Can you share the code that you use for the benchmark ?

hlubek commented:
@dfeyer I think it got lost, basically I used a larger page with many segments and https://github.com/phpbench/phpbench with both implementations

kdambekalns commented:
@dfeyer on that "today" day, did you get to do some testing? Also, @hlubek - do you approve (officially) now?

gerhard-boden commented:
So this reduces performance, yet the advantages outweigh the disadvantages, right?

Would merge within 24 hours if nobody objects.

kitsunet commented:
By now it's even unclear if it really degrades it after I tried to fix a few things
